walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,84 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Set;
 
+/** 
+Algorithm:
+    1. Pick time ranges from events where attendees overlap with meeting request.
+    2. Now find the non overlapping time ranges which have duration more than request duration.
+*/
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    
+    private long meetingDuration;
+    private ArrayList<TimeRange> overlappingRanges;
+    private Collection<String> requestedAttendees;
+    private Collection<TimeRange> requiredRanges;
+    
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        overlappingRanges = new ArrayList<>();
+        
+        requestedAttendees = request.getAttendees();
+        calculateOverlappingRanges(events);
+        
+        meetingDuration = request.getDuration();
+        requiredRanges = new ArrayList<>();
+        getNonOverlappingTimeRanges();
+
+        return requiredRanges;
+    }
+
+    /** 1. Calculate overlapping ranges by iterating in all the events and finding common attendees.*/
+    private void calculateOverlappingRanges(Collection<Event> events) {  
+        for (Event currentEvent: events) {
+            Set<String> currentEventAttendees = currentEvent.getAttendees();
+            boolean overlapRange = false;
+
+            for (String requestedAttendee: requestedAttendees) {
+                if (currentEventAttendees.contains(requestedAttendee)) {
+                    overlapRange = true;
+                    break;
+                }
+            }
+
+            if (overlapRange) {
+                overlappingRanges.add(currentEvent.getWhen());
+            }
+        }
+    }
```

Comment:
can be simplified using java 8 streams. Please refer - https://www.baeldung.com/java-lists-intersection

Reply:
Done.